x-pack/test_serverless/shared/services/user_manager/saml_auth.ts

```diff
@@ -0,0 +1,171 @@
+};
+
+const getCloudUrl = (hostname: string, pathname: string) => {
+  return Url.format({
+    protocol: 'https',
+    hostname,
+    pathname,
+  });
+};
+
+const createCloudSession = async (hostname: string, email: string, password: string) => {
+  const cloudLoginUrl = getCloudUrl(hostname, '/api/v1/users/_login');
+  const sessionResponse: AxiosResponse = await axios.request({
+    url: cloudLoginUrl,
+    method: 'post',
+    data: {
+      email,
+      password,
+    },
+    headers: {
+      accept: 'application/json',
+      'content-type': 'application/json',
+    },
+    validateStatus: () => true,
+    maxRedirects: 0,
+  });
+  const firstName = sessionResponse?.data?.user?.data?.first_name ?? '';
+  const lastName = sessionResponse?.data?.user?.data?.last_name ?? '';
+  const fullname = `${firstName} ${lastName}`.trim();
+  return {
+    token: sessionResponse?.data?.token as string,
+    fullname,
+  };
+};
+
+const createSAMLRequest = async (kbnUrl: string, kbnVersion: string) => {
+  const samlResponse: AxiosResponse = await axios.request({
+    url: kbnUrl + '/internal/security/login',
+    method: 'post',
+    data: {
+      providerType: 'saml',
+      providerName: 'cloud-saml-kibana',
+      currentURL: kbnUrl + '/login?next=%2F"',
+    },
+    headers: {
+      'kbn-version': kbnVersion,
+      'x-elastic-internal-origin': 'Kibana',
+      'content-type': 'application/json',
+    },
+    validateStatus: () => true,
+    maxRedirects: 0,
+  });
+  const cookie = getSessionCookie(samlResponse.headers['set-cookie']![0])!;
+  return { location: samlResponse.data.location, sid: cookie.value };
+};
+
+const createSAMLResponse = async (url: string, ecSession: string) => {
+  const samlResponse = await axios.get(url, {
+    headers: {
+      Cookie: `ec_session=${ecSession}`,
+    },
+  });
+  const $ = cheerio.load(samlResponse.data);
+  const value = $('input').attr('value') ?? '';
+  if (value.length === 0) {
+    throw new Error('Failed to parse SAML response value');
+  }
+  return value;
+};
+
+const finishSAMLHandshake = async ({
+  kbnHost,
+  samlResponse,
+  sid,
+  log,
+}: {
+  kbnHost: string;
+  samlResponse: string;
+  sid?: string;
+  log: ToolingLog;
+}) => {
+  const encodedResponse = encodeURIComponent(samlResponse);
+  let authResponse: AxiosResponse;
+
+  try {
+    authResponse = await axios.request({
+      url: kbnHost + '/api/security/saml/callback',
+      method: 'post',
+      data: `SAMLResponse=${encodedResponse}`,
+      headers: {
+        ...(sid ? { Cookie: `sid=${sid}` } : {}),
+        ...{ 'content-type': 'application/x-www-form-urlencoded' },
+      },
+      validateStatus: () => true,
+      maxRedirects: 0,
+    });
+  } catch (err) {
+    if (axios.isAxiosError(err)) {
+      log.debug(JSON.stringify(err));
+    }
+    throw err;
+  }
+
+  return getSessionCookie(authResponse!.headers['set-cookie']![0])!;
+};
+
+export const createNewSAMLSession = async (params: SAMLSessionParams) => {
```

Comment:
Can we add error handling?  Thrown axios errors can have information we don't want to log in CI.  I'm not sure if there's a catch further up the stack.